### ISSUE TITLE
Allow retrying create process instance in an idempotent way

### ISSUE THREAD

**Issue by npepinpe** [camunda/camunda#12474]
**Is your feature request related to a problem? Please describe.**
There are times where I want to create a process instance with a specific payload only once, or a limited number of times. This is particularly important in SaaS, where I'm billed per root process instance. Sometimes, it can happen that I send a create process instance command, and I don't get a successful response (e.g. time out, response is lost over the network, gateway crashes, etc.), yet the instance is eventually created in Zeebe. Unfortunately, I have no way to know this other than looking into, say, Operate.

In such cases, I don't know if it's safe to retry the request or not; will it create too many process instances?

**Describe the solution you'd like**
I would like a way to reliably know whether I need to retry the request or not - or better, be able to fearlessly retry it until successful, much like we do with deployments, publish messages with IDs, etc.

**Describe alternatives you've considered**
I can use a message start event with a long-ish buffer period, a unique ID, and reuse the unique ID as the correlation key. This works, but is not very intuitive. Additionally, it may not evenly distribute my new process instances the same way the create process instance command would.

**Additional context**

https://jira.camunda.com/browse/SUPPORT-16549

**Comment by Zelldon** [comment]
I agree that the current state is not optimal, I guess one thing which might work as an indication is that when receiving backpressure it was clearly not created and can be retried, on deadline exceeded it is likely to be created we just not received the response in time.

**Comment by korthout** [comment]
This seems like a good improvement. We should consider how long we need to store the information that the create process instance request was recently handled. We cannot store information indefinitely because this would lead to ever increasing state and size on disk.

The message ID is only providing idempotency to messages [while the message is stored in the buffer](https://docs.camunda.io/docs/next/components/concepts/messages/#message-uniqueness). Create Process Instance requests are not stored in a buffer, but of course the process instance is active. So, we could make it idempotent while the process instance is still active (not yet completed/terminated).

If we choose to add this, we should also consider the other not (yet) idempotent RPCs:
- [Broadcast Signal](https://docs.camunda.io/docs/next/apis-tools/grpc/#broadcastsignal-rpc) retrying a broadcast signal would result in new triggers for the signal subscriptions 
- [Create Process Instance With Result](https://docs.camunda.io/docs/next/apis-tools/grpc/#createprocessinstancewithresult-rpc)
- [Modify Process Instance](https://docs.camunda.io/docs/next/apis-tools/grpc/#modifyprocessinstance-rpc) retrying a modification with activate instructions would activate additional tokens

Then there are some RPCs that are not (yet) idempotent, but the impact is less big:
- [Set Variables](https://docs.camunda.io/docs/next/apis-tools/grpc/#setvariables-rpc) depending on how long you wait retrying this may result in overwriting recently updated variables
- [Throw Error](https://docs.camunda.io/docs/next/apis-tools/grpc/#throwerror-rpc) depends a lot on the model but retrying this may result in a new event trigger or a task incident
- [Update Job Retries](https://docs.camunda.io/docs/next/apis-tools/grpc/#updatejobretries-rpc) not idempotent but impact is likely tiny

The other RPCs are idempotent, AFAIK.

**Comment by npepinpe** [comment]
So the issue title is a bit leading. I guess it doesn't have to be idempotent - an alternative would be a way to know if the process instance was indeed created or not.

I discussed with @Zelldon this morning an improvement as well where we could return to clients when the command has been appended to the log, committed to the log, and processed. Processed is the current behavior, i.e. you get the command response. Committed would allow you to know much faster that the command was acknowledged and will definitely be executed eventually. You lose the response, but you know you don't need to retry. 

Of course, we're still stuck with the cases where the client doesn't know if the command was written, or committed. In such cases, being able to retry commands idempotently would be amazing.

**Comment by korthout** [comment]
ZPA triage:
- @aleksander-dytko Please have a look at this problem from a product-management perspective. We see that a solution is useful here, but it's unclear what this should look like at this time.
- @npepinpe Do you see any ways how we could provide a way to know whether the process instance was indeed created or not?

**Comment by aleksander-dytko** [comment]
@korthout recently I was working on https://jira.camunda.com/browse/SUPPORT-15950 where the issue was a bit similar - ensuring idempotency of started process instance. I don't know how much of an issue it is to other customers but maybe we can have a conversation how this might look like from ZPA perspective?

**Comment by ashprojects** [comment]
We also have the same use case. Using CM7's business key solved this perfectly.
If somehow we can restrict multiple process creation using a key.

This becomes critical due to this issue as well: 
https://forum.camunda.io/t/duplicate-process-instances-on-leader-failover/49783

During leader failover sometimes multiple process instances  are created with only one gPRC CreateInstance call.

I don't think Start Message event would solve the actual use case: CreateInstance -> return procesInstanceKey

**Comment by sbuettner** [comment]
Picking up on @korthout regarding other RPCs that could benefit from idempotency. Maybe a general solution similar to https://docs.stripe.com/api/idempotent_requests could be applied across the API.

**Comment by korthout** [comment]
> Maybe a general solution similar to https://docs.stripe.com/api/idempotent_requests could be applied across the API.

That's definitely an interesting idea, and we support it (sort of) for [published messages](https://docs.camunda.io/docs/8.3/components/concepts/messages/#message-uniqueness). I guess the API could be expanded to other RPCs in the same way. The difficulty will be in the implementation, not the API.

> maybe we can have a conversation how this might look like from ZPA perspective?

First, it would require some way to deal with the partitioned data, i.e. requests with the same identifier must always be received by the same partition. Otherwise, we wouldn't be able to detect a request as duplicate. 

Note that we can't detect it in the gateway because only the broker knows whether a request was successfully processed. For example, the response from the broker to the gateway could be lost due to a network failure.

Second, we'd need some way to clean up the data to avoid the system from growing in used disk space indefinitely, for example:
- when the associated entities are removed from Zeebe: would require coupling identified requests and lifecycles of entities
- or after some duration: requires some logic to clear out this data periodically

Third, it may be difficult to resolve this for all requests, as there may be expectations about  load balancing of requests over the partitions. For example, binding request identifiers to partitions creates an additional problem for dynamic scaling of partitions in the same way that message correlation currently forms a problem.

IMO, we should at least await dynamic scaling of partitions to take shape before deciding upon solutions here.

**Comment by pierre-yves-monnet** [comment]
I want to report this ticket from Optum [SUPPORT-20315]( https://jira.camunda.com/browse/SUPPORT-20315) with the same issue.
On high load, the application asks to create a process instance and receive a correct PI number, but two process instances were created behind the scenes.
They detect it because, in each PI, they have a unique business key (and a worker detects that a key shows up two times for two different PIs).

**Comment by nicpuppa** [comment]
We had another related support case [SUPPORT-20941](https://jira.camunda.com/browse/SUPPORT-20941) (internal).

The customer is creating PIs using `"withResult"` option. Under high load they noticed that some PI are duplicated. See this [feature request](https://jira.camunda.com/browse/SUPPORT-20944?jql=%22Customer%20Domain%22%20%3D%20intuit.com) for more information 

cc @aleksander-dytko @megglos

**Comment by megglos** [comment]
> I moved this issue/feature back into the inbox, so we can triage it again

As we didn't comment on moving it to the backlog, we had no indication of a bug and forwarded this to PM for priotisation among other feature requests.

**Comment by sbuettner** [comment]
We discussed this topic again when considering the usage of inbound connectors that call the process creation endpoint. HA setups would benefit from an idempotency solution.
